File: javascript/src/dish.js
const uuid = require('uuid/v4');

function Dish() {
    this.zells = {};
    this.peers = {};
    this.received = {};
}

Dish.prototype.put = function (zell) {
    var id = Math.max(0, Math.max.apply(null, Object.keys(this.zells).map(k => Number.parseInt(k))) + 1);
    this.zells[id] = zell;
    zell.transmit = this.transmit.bind(this);
    return id;
};

Dish.prototype.remove = function (zellId) {
    if (!(zellId in this.zells)) return;
    delete this.zells[zellId];
};

Dish.prototype.join = function (peer) {
    var id = Math.max(0, Math.max.apply(null, Object.keys(this.peers).map(k => Number.parseInt(k))) + 1);

    console.log('join', id);

    this.peers[id] = peer;
    peer.connect().then(() => {
        peer.onSignal((id, signal) => this.receive(id, signal));
        peer.onClose(() => this.leave(id));
    });

    return id;
};

Dish.prototype.leave = function (peerId) {
    if (!(peerId in this.peers)) return;

    console.log('leave', peerId);
    delete this.peers[peerId];
};

Dish.prototype.receive = function (id, signal) {
    var idHex = id.toString('hex');

    if (idHex in this.received) {
        console.log('received', idHex);
        return;
    }
    this.received[idHex] = true;
    console.log('receive', idHex, signal);

    Object.keys(this.zells).forEach(z => this.zells[z].receive(signal));
    Object.keys(this.peers).forEach(p => this.peers[p].sendSignal(id, signal));
};

Dish.prototype.transmit = function (signal) {
    var id = [];
    uuid({}, id);
    this.receive(Buffer.from(id.slice(2, 4)), signal);
};

module.exports = Dish;

File: java/dish/src/main/java/org/zells/dish/Dish.java
package org.zells.dish;

import java.util.HashSet;
import java.util.Set;

public class Dish {
    private Set<Peer> peers = new HashSet<>();
    private Set<Zell> zells = new HashSet<>();
    private Set<Integer> transmitted = new HashSet<>();

    public void put(Zell zell) {
        zells.add(zell);
    }

    public void transmit(int signal) {
        if (transmitted.contains(signal)) {
            return;
        }
        transmitted.add(signal);

        for (Zell z : zells) {
            z.receive(signal);
        }

        for (Peer p : peers) {
            p.write(signal);
        }
    }

    public void join(Peer peer) {
        peers.add(peer);

        new Thread(() -> {
            while(true) {
                int signal = peer.read();
                if (signal == -1) {
                    leave(peer);
                    return;
                }

                transmit(signal);
            }
        }).start();
    }

    public void leave(Peer peer) {
        peers.remove(peer);
        peer.close();
    }
}
